File: app/src/main/java/com/happs/ximand/ringcontrol/model/repository/impl/TrustedDeviceRepository.java
package com.happs.ximand.ringcontrol.model.repository.impl;

import android.app.Application;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteOpenHelper;

import com.happs.ximand.ringcontrol.BuildConfig;
import com.happs.ximand.ringcontrol.model.database.TrustedDevicesDatabaseHelper;
import com.happs.ximand.ringcontrol.model.mapper.Mapper;
import com.happs.ximand.ringcontrol.model.mapper.impl.CursorToTrustedDeviceMapper;
import com.happs.ximand.ringcontrol.model.mapper.impl.TrustedDeviceToContentValuesMapper;
import com.happs.ximand.ringcontrol.model.object.TrustedDevice;

public class TrustedDeviceRepository extends BaseRepository<TrustedDevice> {

    private static TrustedDeviceRepository instance;

    public TrustedDeviceRepository(Application application) {
        super(application);
    }

    public static void initialize(Application application) {
        instance = new TrustedDeviceRepository(application);
    }

    public static TrustedDeviceRepository getInstance() {
        if (instance == null) {
            throw new IllegalStateException("TrustedDevicesRepository was not initialized");
        }
        return instance;
    }

    @Override
    protected SQLiteOpenHelper getSQLiteHelper(Application application) {
        return new TrustedDevicesDatabaseHelper(application, BuildConfig.VERSION_CODE);
    }

    @Override
    protected Mapper<Cursor, TrustedDevice> getFromCursorMapper() {
        return new CursorToTrustedDeviceMapper();
    }

    @Override
    protected Mapper<TrustedDevice, ContentValues> getToContentValuesMapper() {
        return new TrustedDeviceToContentValuesMapper();
    }

    @Override
    protected int getItemId(TrustedDevice item) {
        return item.getId();
    }

    @Override
    protected void setItemId(TrustedDevice item, int id) {
        item.setId(id);
    }
}

File: app/src/main/java/com/happs/ximand/ringcontrol/model/repository/impl/TrustedDeviceRepository.kt
package com.happs.ximand.ringcontrol.model.repository.impl

import android.app.Application
import com.happs.ximand.ringcontrol.BuildConfig
import com.happs.ximand.ringcontrol.model.`object`.TrustedDevice
import com.happs.ximand.ringcontrol.model.database.TrustedDevicesDatabaseHelper
import com.happs.ximand.ringcontrol.model.mapper.impl.CursorToTrustedDeviceMapper
import com.happs.ximand.ringcontrol.model.mapper.impl.TrustedDeviceToContentValuesMapper

class TrustedDeviceRepository(application: Application) : BaseRepository<TrustedDevice>() {

    override val fromCursorMapper = CursorToTrustedDeviceMapper()
    override val toContentValuesMapper = TrustedDeviceToContentValuesMapper()
    override val sqLiteHelper = TrustedDevicesDatabaseHelper(application, BuildConfig.VERSION_CODE)

    companion object {

        private var instance: TrustedDeviceRepository? = null

        fun initialize(application: Application) {
            instance = TrustedDeviceRepository(application)
        }

        fun getInstance(): TrustedDeviceRepository? {
            checkNotNull(instance) { "TrustedDevicesRepository was not initialized" }
            return instance
        }
    }

    override fun getItemId(item: TrustedDevice): Int {
        return item.id
    }

    override fun setItemId(item: TrustedDevice, id: Int) {
        item.id = id
    }
}